File: Piston-master/src/main/java/cc/fyre/piston/listener/FreezeListener.java
package cc.fyre.piston.listener;

import cc.fyre.neutron.NeutronConstants;
import cc.fyre.piston.Piston;
import cc.fyre.piston.PistonConstants;
import cc.fyre.piston.packet.FrozenLogoutPacket;
import cc.fyre.piston.packet.StaffBroadcastPacket;
import cc.fyre.proton.Proton;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.*;

public class FreezeListener implements Listener {

    private static final String FROZEN_MESSAGE = ChatColor.RED + "You cannot do this while frozen.";

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerKick(PlayerKickEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        player.removeMetadata(PistonConstants.FREEZE_METADATA, Piston.getInstance());
        player.sendMessage(FROZEN_MESSAGE);
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onPlayerCommand(PlayerCommandPreprocessEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        String message = event.getMessage();

        // names = {"message", "msg", "tell", "whisper", "t", "w", "m"},
        // names = {"reply", "r"},

        if (message.startsWith("/message")
                || message.startsWith("/msg")
                || message.startsWith("/tell")
                || message.startsWith("/whisper")
                || message.startsWith("/t")
                || message.startsWith("/w")
                || message.startsWith("/m")
                || message.startsWith("/reply")
                || message.startsWith("/r")
        ) {
            return;
        }

        event.setCancelled(true);
    }

    @EventHandler(priority = EventPriority.NORMAL)
    public void onPlayerQuit(PlayerQuitEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        Piston.getInstance().sendPacketAsync(new FrozenLogoutPacket(player.getName()));
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerMove(PlayerMoveEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        final Location to = event.getTo();
        final Location from = event.getFrom();

        if (from.getX() != to.getX() || event.getFrom().getZ() != event.getTo().getZ()) {

            final Location newLocation = from.getBlock().getLocation().add(0.5, 0.0, 0.5);

            newLocation.setPitch(to.getPitch());
            newLocation.setYaw(to.getYaw());

            event.setTo(newLocation);
        }

    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onEntityDamage(EntityDamageEvent event) {

        if (!(event.getEntity() instanceof Player)) {
            return;
        }

        final Player player = (Player) event.getEntity();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onEntityDamageByEntity(EntityDamageByEntityEvent event) {

        if (!(event.getDamager() instanceof Player)) {
            return;
        }

        if (event.getDamager().hasMetadata(PistonConstants.FREEZE_METADATA)) {
            event.setCancelled(true);
            ((Player) event.getDamager()).sendMessage(FROZEN_MESSAGE);
        }

        if (event.getEntity() instanceof Player && event.getEntity().hasMetadata(PistonConstants.FREEZE_METADATA)) {
            ((Player) event.getDamager()).sendMessage(((Player) event.getEntity()).getDisplayName() + ChatColor.RED + " is currently frozen and cannot be damaged.");
        }

    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onInventoryClick(InventoryClickEvent event) {

        final Player player = (Player) event.getWhoClicked();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);
        player.sendMessage(FROZEN_MESSAGE);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerDropItem(PlayerDropItemEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);

        player.updateInventory();
        player.sendMessage(FROZEN_MESSAGE);
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onPlayerInteract(PlayerInteractEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);

        player.updateInventory();
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBlockBreak(BlockBreakEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);

        player.updateInventory();
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBlockPlace(BlockPlaceEvent event) {

        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);

        player.updateInventory();
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onChat(AsyncPlayerChatEvent event) {
        final Player player = event.getPlayer();

        if (!player.hasMetadata(PistonConstants.FREEZE_METADATA)) {
            return;
        }

        event.setCancelled(true);

        String message = ChatColor.translateAlternateColorCodes('&', "&b&l[Frozen] &f" + player.getDisplayName() + "&7: &f" + ChatColor.WHITE + event.getMessage());

        player.sendMessage(message);
        Proton.getInstance().getPidginHandler().sendPacket(new StaffBroadcastPacket(NeutronConstants.STAFF_PERMISSION, message));
    }
}

File: ModSuite-master/bukkit/src/main/kotlin/cc/fyre/modsuite/freeze/listener/FreezeListener.kt
package cc.fyre.modsuite.freeze.listener

import cc.fyre.modsuite.ModSuite
import cc.fyre.modsuite.freeze.FreezeHandler
import cc.fyre.modsuite.mod.ModHandler
import cc.fyre.universe.UniverseAPI
import mkremins.fanciful.FancyMessage
import org.bukkit.ChatColor
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.player.*
import org.bukkit.potion.PotionEffectType

object FreezeListener : Listener {

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onPlayerMove(event: PlayerMoveEvent) {

        if (event.from.x == event.to.x && event.from.z == event.to.z) {
            return
        }

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.teleport(event.from)
        event.player.velocity = FreezeHandler.VELOCITY
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onPlayerCommand(event: PlayerCommandPreprocessEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.sendMessage("${ChatColor.RED}${ChatColor.BOLD}You cannot execute commands whilst frozen!")
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onEntityDamage(event: EntityDamageByEntityEvent) {

        if (event.entity !is Player || event.damager !is Player) {
            return
        }

        val player = event.entity as Player
        val damager = event.damager as Player

        if (FreezeHandler.isFrozen(player)) {
            event.isCancelled = true
            damager.sendMessage("${player.displayName}${ChatColor.RED} is currently frozen.")
            return
        }

        if (!FreezeHandler.isFrozen(damager)) {
            return
        }

        event.isCancelled = true
        damager.sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onInventoryClick(event: InventoryClickEvent) {

        if (event.whoClicked !is Player || !FreezeHandler.isFrozen(event.whoClicked as Player)) {
            return
        }

        event.isCancelled = true
        (event.whoClicked as Player).sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onPlayerDropItem(event: PlayerDropItemEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onPlayerInteract(event: PlayerInteractEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onBlockBreak(event: BlockBreakEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onBlockPlace(event: BlockPlaceEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true
        event.player.sendMessage(FreezeHandler.FROZEN_MESSAGE)
    }

    @EventHandler(priority = EventPriority.NORMAL)
    private fun onPlayerJoin(event: PlayerJoinEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        FreezeHandler.setFrozen(event.player,true)
    }

    @EventHandler(priority = EventPriority.NORMAL)
    private fun onPlayerQuit(event: PlayerQuitEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.player.walkSpeed = 0.2F

        val effects = FreezeHandler.get(event.player) ?: listOf()

        if (effects.isNotEmpty()) {
            event.player.removePotionEffect(PotionEffectType.JUMP)
            event.player.addPotionEffects(effects)
        }

        val message = FancyMessage()
        val displayName = event.player.displayName

        message.then("\n")
        message.then("${displayName}${ChatColor.DARK_RED}${ChatColor.BOLD} has logged out whilst frozen ")
        message.then("on ${UniverseAPI.getServerName()}${ChatColor.DARK_RED}.")
        message.then("\n")
        message.command("ban ${event.player.uniqueId} perm Logged out whilst frozen.")
        message.tooltip("${ChatColor.YELLOW}Click to ban ${displayName}${ChatColor.YELLOW}.")

        ModHandler.getAllOnlineStaffMembers().forEach{message.send(it)}
        /*
        Bukkit.getServer().scheduler.runTaskAsynchronously(ModSuite.instance) {
            ServerModule.sendMessageToNetwork(message,FreezeCommand.PERMISSION,false)
        }*/

        ModSuite.instance.logger.info("${event.player.name} has logged out whilst frozen.")
    }

    @EventHandler(priority = EventPriority.LOWEST)
    private fun onAsyncPlayerChat(event: AsyncPlayerChatEvent) {

        if (!FreezeHandler.isFrozen(event.player)) {
            return
        }

        event.isCancelled = true

        val message = "${ChatColor.RED}${ChatColor.BOLD}[Frozen] ${event.player.displayName}${ChatColor.YELLOW}: ${event.message}"

        ModHandler.getAllOnlineStaffMembers().also{event.player}.forEach{
            it.sendMessage(message)
        }

    }

}